### ISSUE TITLE
Running morphological analyzers in English and Japanese

### ISSUE THREAD

**Issue by notani** [justhalf/bpe_analysis#2]
Normalizing allomorphs (?) so that BPE can find identical morphemes across words.

```
cats -> cat+s
boxes -> box+s

usefulness -> useful+ness
happiness -> happy+ness
```

```
食べる taberu = eat
食べた tabeta = eat+past -> 食べる+た taberu+ta
食べなかった tabenakatta = did not eat -> 食べる+ない+た taberu+nai+ta
```

I expect we get more similar results to UDPipe segmentation if we normalize Japanese morphemes.

**Comment by justhalf** [comment]
Should we keep the plus sign?
In my Indonesian morphology normalizer script I didn't. Also, are you working on this?

**Comment by notani** [comment]
> Should we keep the plus sign?

No, `+` is just for an illustration purpose.

I will do English and Japanese normalization. Were you also working on this?

**Comment by justhalf** [comment]
Yes, the scripts for Indonesian is at https://github.com/justhalf/bpe_analysis/blob/master/morphind/process_txt.py
It uses [MorphInd](http://septinalarasati.com/morphind/) for the morphology analyzer.

> No, `+` is just for an illustration purpose.

I asked because in the Indonesian one I explicitly remove `+`. I think we should remove the plus sign, yeah.

**Comment by notani** [comment]
Did you start English and Japanese normalization, too?

**Comment by justhalf** [comment]
> Did you start English and Japanese normalization, too?

No, I haven't started. I didn't know which morphology analyzer to use. But if we have them, we can simply replace the subprocess call with the corresponding call.

**Comment by justhalf** [comment]
In [this paper](https://www.aclweb.org/anthology/C92-3145) it says the lexicon (19MB) is large:
![image](https://user-images.githubusercontent.com/539150/56667880-98f9db00-667c-11e9-8368-134db54b1f9d.png)

Does anyone know a good English morphology analyzer? I was surprised to find none, only Morfessor, which was automatic.

**Comment by justhalf** [comment]
Maybe this one? http://wiki.apertium.org/wiki/Lttoolbox

**Comment by notani** [comment]
> Maybe this one? http://wiki.apertium.org/wiki/Lttoolbox

Does this output surface forms of morphemes?

This statistical morphological segmenter can generate normalized surface forms like `un+test+able+ly`:
https://github.com/ryancotterell/treeseg

We can search for similar studies by "morphological segmentation" rather than "morphological analysis"

**Comment by justhalf** [comment]
Based on my cursory look, it seems so.

> This statistical morphological segmenter can generate normalized surface forms like un+test+able+ly:
https://github.com/ryancotterell/treeseg

That's a good one, since it is modern. I was looking at more that has more manual analysis, since it will be less automatic, e.g., FST. But couldn't find FST for English.

**Comment by notani** [comment]
How about this?
https://github.com/knowitall/morpha

**Comment by justhalf** [comment]
That looks good. You have one for Japanese as well? (I guess we don't need this for Chinese?)

**Comment by notani** [comment]
Fortunately, Japanese segmentation by UDPipe is already morpheme segmentation and has normalized forms. I think we don't need normalization for Chinese.

Can you do English normalization?

**Comment by justhalf** [comment]
I am trying. Morpha apparently only handles plural nouns and verb inflections, but not derivations. So `happiness` stays as is.
